containerd: ensure overwritten images from load are left dangling

jsternberg opened the pull request:
During a load, ensure any existing images are left as dangling images.
This affects the `docker load` command with the containerd snapshotter
backend.

Fixes the `load` portion of #48907.

```markdown changelog
Fix `docker load` not preserving replaced images
```

vvoland commented on daemon/containerd/image_exporter.go:
```diff
@@ -248,8 +248,21 @@ func (i *ImageService) LoadImage(ctx context.Context, inTar io.ReadCloser, platf
+			}
+
+			// Look up if there is an existing image with this name and ensure a dangling image exists.
+			if img, err := i.resolveImage(ctx, ref.String()); err != nil && !errdefs.IsNotFound(err) {
```
We can simplify `resolveImage` to just `i.images.Get(ctx, ref.String())` because we already have the full reference and don't need to do any fuzzy matching.

vvoland commented on daemon/containerd/image_exporter.go:
```diff
@@ -248,8 +248,21 @@ func (i *ImageService) LoadImage(ctx context.Context, inTar io.ReadCloser, platf
+			// Look up if there is an existing image with this name and ensure a dangling image exists.
+			if img, err := i.resolveImage(ctx, ref.String()); err != nil && !errdefs.IsNotFound(err) {
+				log.G(ctx).WithError(err).Warn("failed to resolve image: %w", err)
+			} else {
+				if err := i.ensureDanglingImage(ctx, img); err != nil {
```
Plain `else` would also match the `if err != nil && errdefs.IsNotFound(err)` case - which will create a dangling image if the image is not found.

(that will be solved by applying the previous suggestion 😅)

jsternberg replied:
I fixed this so it only calls `ensureDanglingImage` when there's no error. I changed the logic so the `else` comes first and the warning message about being unable to find the image happens in the else because it made the logic more clear.

vvoland reviewed (changes requested):
Thanks, that's clever solution!

Could we also have some test for this?

vvoland commented on integration/image/load_test.go:
```diff
@@ -0,0 +1,73 @@
+)
+
+func TestLoadDanglingImages(t *testing.T) {
+	ctx := setupTest(t)
```
Need a skip on Windows (`specialimage.MultiLayerCustom` doesn't support Windows)

jsternberg committed:
containerd: ensure overwritten images from load are left dangling

During a load, ensure any existing images are left as dangling images.
This affects the `docker load` command with the containerd snapshotter
backend.

Signed-off-by: Jonathan A. Sternberg <jonathan.sternberg@docker.com>
